### PR TITLE
[16.0][MIG] connector_ecommerce

### PR DESCRIPTION
**Pull request by juliocabrerizo**
Related issue: https://github.com/OCA/connector-ecommerce/issues/74

Dependencies:
https://github.com/OCA/connector/pull/441

### CODE REVIEW

**Comment by asierneiradev** [comment]
Hi @sbidoul !

Could any maintainer review this proposal for PR?

If any changes are needed, I will adjust it.

Thank you very much

**Comment by asierneiradev** [comment]
Related issue: https://github.com/OCA/connector-ecommerce/issues/74

Dependencies:

- [x] [connector](https://github.com/OCA/connector/pull/442)
- [x] [component_event](https://github.com/OCA/connector/pull/443)
- [x] [connector_base_product](https://github.com/OCA/connector/pull/444)
- [x] [sale_automatic_workflow_payment_mode](https://github.com/OCA/sale-workflow/pull/2321)
- [x] [sale_exception](https://github.com/OCA/sale-workflow/pull/2344)

**Comment by asierneiradev** [comment]
Good afternoon @simahawk !

Do you consider that it is ready to be merge?

Thank you very much

File: connector_ecommerce/migrations/14.0.1.0.1/post-migration.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+from openupgradelib import openupgrade
```
Comment (RodrigoBM) [inline connector_ecommerce/migrations/14.0.1.0.1/post-migration.py]: this directory must be deleted

File: connector_ecommerce/models/invoice.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+        return res
+
+    def action_invoice_paid(self):
+        res = super().action_invoice_paid()
+        for record in self:
+            self._event("on_invoice_paid").notify(record)
+        return res
+
+    def notify_invoice_validate(self):
+        for record in self.filtered(lambda m: m.move_type == "out_invoice"):
```
Comment (RodrigoBM) [inline connector_ecommerce/models/invoice.py]: I would change this to
```
for record in self:
    if move_type != "out_invoice":
        continue
```

File: connector_ecommerce/models/product.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+                self._event("on_product_price_changed").notify(product)
+
+    def write(self, vals):
+        self_context = self.with_context(from_product_ids=self.ids)
+        result = super(ProductProduct, self_context).write(vals)
+        self._price_changed(vals)
+        return result
+
+    @api.model
+    def create(self, vals):
```
Comment (RodrigoBM) [inline connector_ecommerce/models/product.py]: 2023-03-10 15:28:49,600 1 WARNING gl_433 odoo.api.create: The model odoo.addons.connector_ecommerce.models.product is not overriding the create method in batch

File: connector_ecommerce/models/sale.py [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+                )
+                % order.name
+            )
+            for picking in order.picking_ids:
+                picking.message_post(body=message)
+            for invoice in order.invoice_ids:
+                invoice.message_post(body=message)
+
+    @api.model
+    def create(self, values):
```
Comment (RodrigoBM) [inline connector_ecommerce/models/sale.py]: 2023-03-10 15:28:49,601 1 WARNING gl_433 odoo.api.create: The model odoo.addons.connector_ecommerce.models.sale is not overriding the create method in batch

File: connector_ecommerce/models/product.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+            # to fire the event 2 times
+            if self.env.context.get("from_product_ids"):
+                from_product_ids = self.env.context["from_product_ids"]
+                remove_products = product_model.browse(from_product_ids)
+                products -= remove_products
+            for product in products:
+                self._event("on_product_price_changed").notify(product)
+
+    def write(self, vals):
+        result = super(ProductTemplate, self).write(vals)
```
Comment (RodrigoBM) [inline connector_ecommerce/models/product.py]: Apply PEP 3135 (28-Apr-2007) to "super()."

File: connector_ecommerce/models/product.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+    def write(self, vals):
+        self_context = self.with_context(from_product_ids=self.ids)
+        result = super(ProductProduct, self_context).write(vals)
+        self._price_changed(vals)
+        return result
+
+    @api.model
+    def create(self, vals):
+        product = super(ProductProduct, self).create(vals)
```
Comment (RodrigoBM) [inline connector_ecommerce/models/product.py]: Apply PEP 3135 (28-Apr-2007) to "super()."

File: connector_ecommerce/models/sale.py [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+        order = super(SaleOrder, self).create(values)
```
Comment (RodrigoBM) [inline connector_ecommerce/models/sale.py]: Apply PEP 3135 (28-Apr-2007) to "super()."

File: connector_ecommerce/models/sale.py [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+        if values.get("canceled_in_backend"):
+            order._log_canceled_in_backend()
+            order._try_auto_cancel()
+        return order
+
+    def write(self, values):
+        result = super(SaleOrder, self).write(values)
```
Comment (RodrigoBM) [inline connector_ecommerce/models/sale.py]: Apply PEP 3135 (28-Apr-2007) to "super()."

File: connector_ecommerce/models/sale.py [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+        if values.get("canceled_in_backend"):
+            self._log_canceled_in_backend()
+            self._try_auto_cancel()
+        return result
+
+    def action_cancel(self):
+        res = super(SaleOrder, self).action_cancel()
```
Comment (RodrigoBM) [inline connector_ecommerce/models/sale.py]: Apply PEP 3135 (28-Apr-2007) to "super()."

File: connector_ecommerce/models/sale.py [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+        action["res_id"] = parent.id
+        return action
+
+    def _create_delivery_line(self, carrier, price_unit):
+        if self.order_line.filtered(lambda r: r.is_delivery):
+            # skip if we have already a delivery line (created by
+            # import of order)
+            return
+        else:
+            return super(SaleOrder, self)._create_delivery_line(carrier, price_unit)
```
Comment (RodrigoBM) [inline connector_ecommerce/models/sale.py]: Apply PEP 3135 (28-Apr-2007) to "super()."

File: connector_ecommerce/models/stock.py [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+    _inherit = "stock.picking"
+
+    related_backorder_ids = fields.One2many(
+        comodel_name="stock.picking",
+        inverse_name="backorder_id",
+        string="Related backorders",
+    )
+
+    def write(self, vals):
+        res = super(StockPicking, self).write(vals)
```
Comment (RodrigoBM) [inline connector_ecommerce/models/stock.py]: Apply PEP 3135 (28-Apr-2007) to "super()."

File: connector_ecommerce/tests/test_invoice_event.py [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+from unittest import mock
+
+import odoo.tests.common as common
+
+
+class TestInvoiceEvent(common.TransactionCase):
+    """Test if the events on the invoice are fired correctly"""
+
+    def setUp(self):
+        super(TestInvoiceEvent, self).setUp()
```
Comment (RodrigoBM) [inline connector_ecommerce/tests/test_invoice_event.py]: Apply PEP 3135 (28-Apr-2007) to "super()."

File: connector_ecommerce/tests/test_onchange.py [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+
+DB = common.DB
+ADMIN_USER_ID = common.ADMIN_USER_ID
+
+
+class TestOnchange(ComponentRegistry):
+    """Test if the onchanges are applied correctly on a sales order"""
+
+    def setUp(self):
+        super(TestOnchange, self).setUp()
```
Comment (RodrigoBM) [inline connector_ecommerce/tests/test_onchange.py]: Apply PEP 3135 (28-Apr-2007) to "super()."

File: connector_ecommerce/tests/test_picking_event.py [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+        }
+        vals.update(**values)
+        pack_operation = move_line_env.new(vals)
+        pack_operation.onchange_product_id()
+        return move_line_env.create(
+            pack_operation._convert_to_write(pack_operation._cache)
+        )
+
+    def setUp(self):
+        super(TestPickingEvent, self).setUp()
```
Comment (RodrigoBM) [inline connector_ecommerce/tests/test_picking_event.py]: Apply PEP 3135 (28-Apr-2007) to "super()."

File: connector_ecommerce/components/line_builder.py [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+        builder.get_line()
+
+    """
+
+    _name = "ecommerce.order.line.builder"
+    _inherit = "base.connector"
+    _usage = "order.line.builder"
+
+    def __init__(self, work_context):
+        super(SpecialOrderLineBuilder, self).__init__(work_context)
```
Comment (RodrigoBM) [inline connector_ecommerce/components/line_builder.py]: Apply PEP 3135 (28-Apr-2007) to "super()."

File: connector_ecommerce/components/line_builder.py [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+
+class ShippingLineBuilder(Component):
+    """Return values for a Shipping line"""
+
+    _name = "ecommerce.order.line.builder.shipping"
+    _inherit = "ecommerce.order.line.builder"
+    _usage = "order.line.builder.shipping"
+
+    def __init__(self, work_context):
+        super(ShippingLineBuilder, self).__init__(work_context)
```
Comment (RodrigoBM) [inline connector_ecommerce/components/line_builder.py]: Apply PEP 3135 (28-Apr-2007) to "super()."

File: connector_ecommerce/components/line_builder.py [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+        self.product_ref = ("connector_ecommerce", "product_product_shipping")
+        self.sequence = 999
+
+    def get_line(self):
+        values = super(ShippingLineBuilder, self).get_line()
```
Comment (RodrigoBM) [inline connector_ecommerce/components/line_builder.py]: Apply PEP 3135 (28-Apr-2007) to "super()."

File: connector_ecommerce/components/line_builder.py [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+
+class CashOnDeliveryLineBuilder(Component):
+    """Return values for a Cash on Delivery line"""
+
+    _name = "ecommerce.order.line.builder.cod"
+    _inherit = "ecommerce.order.line.builder"
+    _usage = "order.line.builder.cod"
+
+    def __init__(self, work_context):
+        super(CashOnDeliveryLineBuilder, self).__init__(work_context)
```
Comment (RodrigoBM) [inline connector_ecommerce/components/line_builder.py]: Apply PEP 3135 (28-Apr-2007) to "super()."

File: connector_ecommerce/components/line_builder.py [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+
+class GiftOrderLineBuilder(Component):
+    """Return values for a Gift line"""
+
+    _name = "ecommerce.order.line.builder.gift"
+    _inherit = "ecommerce.order.line.builder"
+    _usage = "order.line.builder.gift"
+
+    def __init__(self, work_context):
+        super(GiftOrderLineBuilder, self).__init__(work_context)
```
Comment (RodrigoBM) [inline connector_ecommerce/components/line_builder.py]: Apply PEP 3135 (28-Apr-2007) to "super()."

File: connector_ecommerce/components/line_builder.py [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+        self.product_ref = ("connector_ecommerce", "product_product_gift")
+        self.sign = -1
+        self.gift_code = None
+        self.sequence = 990
+
+    def get_line(self):
+        line = super(GiftOrderLineBuilder, self).get_line()
```
Comment (RodrigoBM) [inline connector_ecommerce/components/line_builder.py]: Apply PEP 3135 (28-Apr-2007) to "super()."

File: connector_ecommerce/models/product.py [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+
+class ProductTemplate(models.Model):
+    _inherit = "product.template"
+
+    # TODO implement set function and also support multi tax
+    @api.depends("taxes_id.tax_group_id")
+    def _compute_tax_group_id(self):
+        self.ensure_one()
+        taxes = self.taxes_id
+        self.tax_group_id = taxes[:-1].tax_group_id.id
```
Comment (RodrigoBM) [inline connector_ecommerce/models/product.py]: I think this is not correct, if you can some taxes_id and you make taxes[:-1] get more result.

```
>>> list_element = [1,2,3,4]
>>> list_element[:-1]
[1, 2, 3]
```
and if after you get tax_group_id some element you can get singleton.

```
>>> self.env["account.tax"].search([]).tax_group_id.id
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/mnt/environment/odoo-server/odoo/fields.py", line 5035, in __get__
    raise ValueError("Expected singleton: %s" % record)
ValueError: Expected singleton: account.tax.group(1, 2)

```

I think the right thing to do is 
taxes[-1:].tax_group_id.id

```
>>> list_element = [1,2,3,4]
>>> list_element[-1:]
[4]
>>> list_element = []
>>> list_element[-1:]
[]

```

**Review by RodrigoBM — CHANGES_REQUESTED** [review]

File: connector_ecommerce/models/product.py [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+    tax_group_id = fields.Many2one(
+        comodel_name="account.tax.group",
+        compute="_compute_tax_group_id",
+        string="Tax Group",
+        help="Tax groups are used with some external system like Prestashop",
+    )
+
+    @api.model
+    def _price_changed_fields(self):
+        return {"list_price", "lst_price", "standard_price"}
```
Comment (RodrigoBM) [inline connector_ecommerce/models/product.py:27]: lst_price does not exist in product template

Comment (RodrigoBM) [inline connector_ecommerce/models/product.py:27]: lst_price does not exist in product template

File: connector_ecommerce/models/product.py [hunk 23]
```diff
@@ -0,0 +1,90 @@
...
+        self._price_changed(vals)
+        return result
+
+
+class ProductProduct(models.Model):
+    _inherit = "product.product"
+
+    @api.model
+    def _price_changed_fields(self):
+        return {"lst_price", "standard_price", "price", "price_extra"}
```
Comment (RodrigoBM) [inline connector_ecommerce/models/product.py]: price field dead in V16 _May he rest in peace_

I would change the return to

```
    @api.model
    def _price_changed_fields(self):
        return {
            "lst_price",
            "price_extra",
        } | self.product_tmpl_id._price_changed_fields()
```

**Comment by asierneiradev** [comment]
Hi @RodrigoBM !

Thank you very much for the suggested changes.

Could you check again the PR proposal ?

Regards

**Review by RodrigoBM — APPROVED** [review]

**Review by RodrigoBM — APPROVED** [review]

**Comment by john-herholz-dt** [comment]
@RodrigoBM can you merge this?

**Review by john-herholz-dt — APPROVED** [review]

**Comment by RodrigoBM** [comment]
@john-herholz-dt I wish I could, but I have not yet been blessed with that power.

**Comment by john-herholz-dt** [comment]
@lmignon can you merge this PR?

**Comment by john-herholz-dt** [comment]
Thanks @lmignon you helped a lot with this. I can finally continue using this module for production.

**Pull request merged**